app/controllers/admin/passwords_controller.rb

```diff
@@ -3,4 +3,44 @@
 class Admin::PasswordsController < Devise::PasswordsController
   before_action :user_logout_required
   layout "session"
+
+  before_action :check_if_totp_required, only: [:edit]
+  before_action :verify_otp_code, only: [:update]
+
+  def check_if_totp_required
+    admin = find_admin_by_reset_password_token(params[:reset_password_token])
+
+    return unless admin
+
+    @totp_required = admin.otp_required_for_login
+  end
+
+  def verify_otp_code
+    admin = find_admin_by_reset_password_token(admin_params[:reset_password_token])
+
+    return unless admin
+
+    return unless admin.otp_required_for_login && !valid_otp_attempt?(admin)
+
+    flash[:error] = t("admin.sessions.invalid_totp")
+
+    redirect_to edit_admin_password_path(reset_password_token: admin_params[:reset_password_token])
```

Comment:
I think this should `render` instead of redirect, since we have a form fill failure and likely want to keep the input. If yes, then the flash should be a `flash.now`